CASMINST-5480: Correct problems with NCN personalization procedure for standard upgrade case (#2779)
Cray-HPE/docs-csm: operations/configuration_management/NCN_Node_Personalization.md

configuration will only be applied to appropriate node types.

## Prerequisites

- The Cray CLI must be configured and authenticated on the node where this procedure is being run.
  - See [Configure the Cray CLI](../configure_cray_cli.md).
- The latest CSM documentation RPM must be installed on the node where this procedure is being run.
  - See [Check for latest documentation](../../update_product_stream/README.md#check-for-latest-documentation).

## Procedure

The following procedure creates a new CFS configuration using the default `sat bootprep` files from the `hpc-csm-software-recipe` repository in VCS.

### 1. Preparation

1. (`ncn-m#`) Back up the current CFS state.

    For the CSM 1.3 release, the NCN personalization configuration is named `ncn-personalization`.
    If a CFS configuration with this name already exists, the procedure on this page will overwrite it.

    Because this procedure will create/update a CFS configuration and modify the CFS components of the management NCNs,
    take a snapshot of the current state of these CFS objects.

    ```bash
    /usr/share/doc/csm/scripts/operations/configuration/backup_cfs_config_comp.sh
    ```

> 1. (`ncn-m#`) Perform the steps in the [Accessing `sat bootprep` Files](Accessing_Sat_Bootprep_Files.md) procedure to gather a copy of the `sat bootprep` files.

Author: ```suggestion
1. (`ncn-m#`) Gather a copy of the `sat bootprep` files.

    See [Accessing `sat bootprep` Files](Accessing_Sat_Bootprep_Files.md).
```

# NCN Node Personalization

- [Background](#background)
- [Prerequisites](#prerequisites)
- [Procedure](#procedure)
  1. [Preparation](#1-preparation)
  1. [Edit CPE and Analytics layers](#2-edit-cpe-and-analytics-layers)
  1. [Disable CFS on management NCNs](#3-disable-cfs-on-management-ncns)
  1. [Update CFS configuration and components](#4-update-cfs-configuration-and-components)
  1. [Cleanup](#5-cleanup)

## Background

> NCN node personalization refers to the process of CFS applying a configuration to a management NCN after it is booted.
> During CSM installs and upgrades, this CFS configuration must be created and set as the desired configuration in CFS for all management NCNs.
> The purpose is to ensure that the appropriate CFS layers are applied to the management NCNs after they boot.

Author: ```suggestion
NCN node personalization refers to the process of CFS applying a configuration to a management NCN after it is booted.
During CSM installs and upgrades[^1], this CFS configuration must be created and set as the desired configuration in CFS for all management NCNs.
The purpose is to ensure that the appropriate CFS layers are applied to the management NCNs after they boot.

[^1]: Except for CSM-only installs and upgrades, which are very rare in production environments. If CSM is the only product
on the system, then this procedure is not performed.
```

> The same CFS configuration is used for post-boot personalization of master, storage, and worker NCNs. However, some individual parts of that

Reviewer: is there a reason for this line break between line 18 and 19?

Author: Line length. Like HTML, when the markdown is rendered, the line break goes away. So our source files usually start new lines somewhere after 60-80-ish characters.

    - name: ncn-personalization
    ```

### 2. Edit CPE and Analytics layers

> This section is required if this procedure is being performed as part of a CSM upgrade.

Reviewer 2: I feel like this section wouldn't be required if this procedure was being performed as part of a CSM*-only* upgrade, I think when I wrote "in the case of an upgrade" I was thinking more like a full recipe upgrade.

Author: For sure. The CSM upgrade docs make that clear -- this page is only linked from the section that is for what we call the "standard upgrade", which is the non-CSM-only case. That said, it wouldn't hurt for me to be clearer here as well, to reduce the possibility of confusion. Good catch.

If that is not the case, then skip ahead to [Update CFS configuration and components](#4-update-cfs-configuration-and-components).

In this section, the `management-bootprep-node-personalization.yaml` file is modified to specify the versions of CPE and Analytics that
are currently in use on the system. This is necessary because the new versions of CPE and Analytics have not yet been installed at this
time in the CSM upgrade procedure.

1. (`ncn-m#`) Acquire a copy of the current CPE and Analytics products CFS configuration values already in use.

>     There are multiple ways to do this. The following is an example of finding these values by looking at the CFS configuration of a worker NCN on the system.

Reviewer 2: Do you have any concern here about the potential for some NCNs to have different configurations?

Not saying it has to be done with SAT and as you say it can be done different ways, but using `sat status` here would tell you the desired configurations for all the NCNs at once.

```
ncn-m001:~ # sat status --filter role=management --fields xname,desiredconfig
+---------------+---------------------+
| xname         | Desired Config      |
+---------------+---------------------+
| x3000c0s1b0n0 | ncn-personalization |
| x3000c0s2b0n0 | ncn-personalization |
| x3000c0s3b0n0 | ncn-personalization |
| x3000c0s4b0n0 | ncn-personalization |
| x3000c0s5b0n0 | ncn-personalization |
| x3000c0s6b0n0 | ncn-personalization |
| x3000c0s7b0n0 | ncn-personalization |
| x3000c0s8b0n0 | ncn-personalization |
| x3000c0s9b0n0 | ncn-personalization |
+---------------+---------------------+
```

And if you wanted to be really fancy:
```
ncn-m001:~ # sat status --filter role=management --fields desiredconfig --format json | jq -r '.[]."Desired Config"' | uniq
ncn-personalization
```

Scrolling down in the procedure a little I see that it doesn't REALLY matter that much because the procedure will have you update the `desiredConfig` for all the nodes at once to this config anyway, so by the end they will all have the same config.

Author: Using sat for this definitely seems simpler, so I'll make that edit. And if they are using multiple configurations, then they'd probably want to know that so they could make sure that they are using the one they wanted as the source. Good call.

    1. Find the component name (xname) of one of the NCN worker nodes.

        ```bash
        WORKER_XNAME=$(cray hsm state components list --role Management --type Node --subrole Worker --format json | jq -r '.Components[0].ID')
        echo "${WORKER_XNAME}"
        ```

        Example output:

        ```text
        x3000c0s7b0n0
        ```

    1. Find the current desired CFS configuration for this worker.

        ```bash
        WORKER_CONFIG=$(cray cfs components describe "${WORKER_XNAME}" --format json | jq -r '.desiredConfig')
        echo "${WORKER_CONFIG}"
        ```

        Example output:

        ```text
        ncn-personalization
        ```

    1. Show the CPE and Analytics layers of that configuration.

        ```bash

>         cray cfs configurations describe "${WORKER_CONFIG}" --format json | \

Reviewer 2: I'm totally good with it and you are probably aware, but I just wanted to point out that this is different than what we had - we were looking at `cray cfs components describe` in order to see what configuration had last successfully run for the component, whereas this is looking at the desired config layers for the node.

If anything I think this might be better to look at a configuration rather than the state of an individual component.

I don't recall whether we talked any more about the pros and cons, the closest I can find is this slack thread: https://cray.slack.com/archives/CLG4R6SU9/p1665604546333739

Author: Yeah, I don't think we discussed it, but my line of thought when I came to that part was exactly like you laid out here. I figured that the configuration was a better source of truth. I had meant to ask you about that but forgot, so I'm glad you spotted it.

        ```

    1. In the output of the previous step, note the values for `cloneUrl`, `commit`, and `playbook` for both layers.

1. (`ncn-m#`) Update the `management-bootprep-node-personalization.yaml` file with the CPE and Analytics values identified in the previous step.

    In order to accurately represent the exact configuration already in use for CPE and Analytics, use the `git` key in the `bootprep` layer definition.
    For each of these two layers:

    - Ensure that `commit` and `playbook` match the values identified in the previous step.
    - Ensure that `url` matches the `cloneUrl` value identified in the previous step.

    The example below shows what the CPE and Analytics layers should look like, based on the earlier example output.

    ```yaml
    - name: cpe-pe_deploy-integration-{{cpe.version}}
      playbook: pe_deploy.yml
      git:
        url: https://api-gw-service-nmn.local/vcs/cray/cpe-config-management.git
        commit: c38c16bc9a645a9339a8303865e696b089b15e17
    - name: analytics-site-integration-{{analytics.version}}
      playbook: site.yml
      git:
        url: https://api-gw-service-nmn.local/vcs/cray/analytics-config-management.git

>         commit: c9c0b2cc69998830a47d7c989b07f550814af095

Reviewer: Do we want to remove the `{{cpe.version}}` and `{{analytics.version}}` above?   If we leave this as is, the name will be the papaya version but the commit id will be the raspberry version.

Author: Good call.
